### DISCUSSION TITLE
If an error occurs on the build phase, how to know cause?

### DISCUSSION THREAD

**Discussion by jiyong00** [pyuvm/pyuvm#149] [Q&A]
If I sometimes use the wrong variable name or class name in the build phase, an error occurs. At this time, the following NullTrigger error occurs, so I don't know where the problem is. How are other guys solving it?

     0.00ns INFO     cocotb                             UVM running test simple_test
Traceback (most recent call last):
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/scheduler.py", line 633, in _resume_coro_upon
    trigger.prime(self._react)
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/triggers.py", line 685, in prime
    callback(self)
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/scheduler.py", line 384, in _react
    self._event_loop(trigger)
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/scheduler.py", line 487, in _event_loop
    self._schedule(coro, trigger=trigger)
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/scheduler.py", line 957, in _schedule
    self._unschedule(coroutine)
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/scheduler.py", line 571, in _unschedule
    self._cleanup()
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/scheduler.py", line 1094, in _cleanup
    assert not self._pending_coros
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../cocotb/lib/python3.9/site-packages/cocotb/__init__.py", line 233, in _initialise_testbench
    _initialise_testbench_(argv_)
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/__init__.py", line 323, in _initialise_testbench_
    regression_manager._execute()
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/regression.py", line 554, in _execute
    return self._start_test()
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/regression.py", line 576, in _start_test
    cocotb.scheduler._add_test(self._test_task)
  File "/...cocotb/lib/python3.9/site-packages/cocotb/scheduler.py", line 842, in _add_test
    self._resume_coro_upon(
  File "/.../cocotb/lib/python3.9/site-packages/cocotb/scheduler.py", line 636, in _resume_coro_upon
    self._trigger2coros.pop(trigger)
KeyError: <NullTrigger for Start <Test run_test> at 0x7f9016304cc0>
     0.00ns ERROR    cocotb.scheduler                   Failing test at simulator request before test run completion: Simulator shut down prematurely
sys:1: RuntimeWarning: coroutine 'Reset.start' was never awaited
xmsim: *W,RNQUIE: Simulation is complete.

**Comment by raysalemi** [comment]
What makes you say that this is being caused by a build_phase error?

It is hard to debug from the error message alone.  Can you share some of your code to demonstrate what you mean?

**Reply by jiyong00** [reply]
Unfortunately, I cannot share the source code. However, when I try to use a variable or class that does not exist in the build phase, of course an error occurs, but there are cases where the error log does not show which code actually occurs.

**Reply by raysalemi** [reply]
> Unfortunately, I cannot share the source code. However, when I try to use a variable or class that does not exist in the build phase, of course an error occurs, but there are cases where the error log does not show which code actually occurs.

Can you run the code without starting the simulator? Then Python could provide the dereference error.

**Reply by jiyong00** [reply]
I'll share some of the code.

```
class Reset:
    def __init__(self, reset, clock, cycle=1000, active_low=False):
        self.reset = reset
        self.clock = clock
        self.cycle = cycle
        self.active_low = active_low

    async def start(self):
        '''Reset the DUT'''
        cocotb.log.info('Resetting DUT')
        self.reset.value = 1 if self.active_low is False else 0
        for _ in range(self.cycle):
            await FallingEdge(self.clock)
        cocotb.log.info('Out of reset')
        self.reset.value = 0 if self.active_low is False else 1

@cocotb.test()
async def run_test(dut):
    cocotb.log.info(f"Seeding Python random module with {hex(cocotb.RANDOM_SEED)}")

    testname = cocotb.plusargs['UVM_TESTNAME']
    cocotb.log.info(f"UVM running test {testname}")

    clock = Clock(dut.clock, 20, units='ns')
    cocotb.start_soon(clock.start())

    reset = Reset(dut.reset, dut.clock, cycle=10)
    cocotb.start_soon(reset.start())

    await uvm_root().run_test(testname)
```



If I delete "cocotb.start_soon(reset.start())" here, the error is reported without problems.
So it seems that this code blocked error reporting in the build phase.

**Reply by jiyong00** [reply]
So, I think clock and reset control should be moved to run_phase.

**Reply by raysalemi** [reply]
> So, I think clock and reset control should be moved to run_phase.

Everything that can block must be in `run_phase`

Also, check out the `@pyuvm.test()` decorator.

**Comment by timothyscherer** [comment]
The issue is that you did not await `cocotb.start_soon(reset.start())`, hence the message `sys:1: RuntimeWarning: coroutine 'Reset.start' was never awaited`.

And yes, it is true that anything using async / await must be called from the build phase. You cannot call an async from a non-async function, and build_phase is the only async phase of pyuvm.
